pkg/controller/util/endpoint/controller_utils.go

```diff
@@ -164,27 +164,39 @@ func PodChanged(oldPod, newPod *v1.Pod, endpointChanged EndpointsMatch) (bool, b
 		labelsChanged = true
 	}
 
-	// If the pod's deletion timestamp is set, remove endpoint from ready address.
-	if newPod.DeletionTimestamp != oldPod.DeletionTimestamp {
+	// If the pod's deletion timestamp has been set, remove endpoint from ready address.
+	if oldPod.DeletionTimestamp == nil && newPod.DeletionTimestamp != nil {
 		return true, labelsChanged
 	}
+
 	// If the pod's readiness has changed, the associated endpoint address
 	// will move from the unready endpoints set to the ready endpoints.
 	// So for the purposes of an endpoint, a readiness change on a pod
 	// means we have a changed pod.
 	if podutil.IsPodReady(oldPod) != podutil.IsPodReady(newPod) {
 		return true, labelsChanged
 	}
-	// Convert the pod to an Endpoint, clear inert fields,
-	// and see if they are the same.
-	// TODO: Add a watcher for node changes separate from this
-	// We don't want to trigger multiple syncs at a pod level when a node changes
-	return endpointChanged(newPod, oldPod), labelsChanged
+
+	// Check if the pod IPs have changed
+	if len(oldPod.Status.PodIPs) != len(newPod.Status.PodIPs) {
+		return true, labelsChanged
+	}
+	for i := range oldPod.Status.PodIPs {
+		if oldPod.Status.PodIPs[i].IP != newPod.Status.PodIPs[i].IP {
+			return true, labelsChanged
+		}
+	}
```

Comment:
This has the unfortunate consequence of triggering a resync for a IPv4 Service if a Pod's IPv6 address changes. Is there anyway we could avoid that without having access to the Service? I'm thinking this would be a relatively rare occurrence so probably not be a big deal.

Reply:
indeed it should not be a normal behavior that a pod  changes only one IP, maybe the pod restarts and acquires the same for one family and not for the other? nevertheless it seems safer to resync than trying to optimize